### ISSUE TITLE
Trouble with Redirects and Sitemap

### ISSUE THREAD

**Issue by yesezra** [middleman/middleman#1166]
Hello,

I'm using the [redirects feature](https://github.com/middleman/middleman/blob/v3-stable/middleman-core/lib/middleman-core/sitemap/extensions/redirects.rb) to maintain paths when my Middleman website changes. Unfortunately, I've been experiencing some issues in the development environment.

I made a [test case](https://github.com/ahhrrr/middleman-redirect-test-case) to illustrate my problem. In `config.rb`, I have a redirect set up like so:

```
redirect 'redirect_me/index.html', to: 'destination.html'
```

The redirect seems to work as expected, but when I browse to the Sitemap web view, I get the following error in the console:

```
NoMethodError: super: no superclass method `get_source_file' for #<Middleman::Sitemap::Extensions::Redirects::RedirectResource:0x007fb10ea7d868>
```

And here's what I see in my browser:
![screen shot 2014-02-03 at 1 52 32 pm](https://f.cloud.github.com/assets/173440/2070586/826b9e98-8d1d-11e3-817b-27c776071f59.png)

Am I configuring the redirect incorrectly, or is this a bug? 

Also, I haven't been able to find any documentation, other than the Cucumber feature, on the redirect feature. Is it fully supported? Can I help with documentation?

Thanks!

**Comment by rowanhogan** [comment]
I was having this same issue, though it was when I was generating an XML sitemap (and was therefore preventing `middleman build` from succeeding).

I was able to solve that issue by excluding the redirect resources when printing them in the sitemap with the following method:

```

**Comment by tdreyno** [comment]
would `r.is_a? Redirects::RedirectResource` work? That seems reasonable to me

**Comment by rowanhogan** [comment]
Good point @tdreyno, the full class name with namespaces is `Middleman::Sitemap::Extensions::Redirects::RedirectResource`, so my method becomes:

```
resources = sitemap.resources.reject { |r| r.is_a? Middleman::Sitemap::Extensions::Redirects::RedirectResource }
```

**Comment by calebhailey** [comment]
Did this ever get fixed? I just ran into this today using Middleman 3.2.1 on Ruby 2.1.2. 

Thanks in advance for your time and for Middleman!

**Comment by milesmatthias** [comment]
Not sure if this is related or not, but I see redirects served up as plain text, not html, so the redirect doesn't happen on the client.

**Comment by tdreyno** [comment]
@milesmatthias In the Middleman preview server or on your deployed server?

**Comment by tdreyno** [comment]
It got fixed, but maybe not back ported to v3. @calebhailey is your issue the original (in the MM debug view?) or something from later in the thread?

**Comment by milesmatthias** [comment]
@tdreyno both. I'm using v3.4.0.

**Comment by tdreyno** [comment]
May not help, but give `v3.4.1` a whirl?

**Comment by matiasgarciaisaia** [comment]
Hi there!

I've hit this very same issue while developing our website.

I'm trying to mimic our `/projects/` tree with an alias `/work/`. So, if we have `/projects/acme-inc/` and `/projects/secret-stuff-x/`, going to `/work/acme-inc/` or `/work/secret-stuff-x/` redirects you to the `/projects/` pages.

The relevant code in `config.rb` file was like this:

```ruby
ready do
  project_pages = sitemap.resources.select { |resource| resource.destination_path.start_with? 'projects/' }
  project_pages.each do |page|
    destination_path = page.destination_path
    redirect (destination_path.sub "projects/", "work/"), to: destination_path
  end
end
```

Like... filter all the `/projects/` URLs, then make a redirect from it's `/work/` counterpart to it.

As `RedirectResource` wasn't defining a `get_source_file` method, I've reopened the class to define it as the target's `source_file`:

```ruby
class Middleman::Sitemap::Extensions::Redirects::RedirectResource
  def get_source_file
    @request_path.source_file
  end
end
```

And, instead of passing the path in `redirect`'s `to:` value, I now pass the actual middleman resource:

```ruby
ready do
  project_pages = sitemap.resources.select { |resource| resource.destination_path.start_with? 'projects/' }
  project_pages.each do |page|
    destination_path = page.destination_path
    redirect (destination_path.sub "projects/", "work/"), to: page # <-- page is a Middleman::Sitemap::Resource
  end
end
```

If you have a path and want to get it's `Resource`, you can query the sitemap with `sitemap.find_resource_by_path` or `sitemap.find_resource_by_destination_path`.

Hope this helps avoid some headaches!

**Comment by sandstrom** [comment]
I'm doing some issue-gardening 🌿🌷🌾 and came across this issue.

@matiasgarciaisaia is this an issue with Middleman? If so, would you be open to making a PR with a fix + test? That would be an awesome thing! 😎

**Comment by matiasgarciaisaia** [comment]
@sandstrom I wanted to test the changes, but really couldn't find how to. I've opened #2006 with the change, at least.

**Comment by sandstrom** [comment]
@matiasgarciaisaia Awesome! ⛵️  I've commented in that issue.

**Comment by matiasgarciaisaia** [comment]
For anyone that comes by: this issue should be fixed starting in v3.3.7. Read #2006's comments for more info.

I'm not sure why @milesmatthias has seen this in v3.4.0 - are you sure about that?

I'd totally close the issue, otherwise.
